### ISSUE TITLE
blockchain555 - OptionTokens corresponding to the dust amount of paymentToken cannot be exercised.

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-06-velocimeter-judging#176]
blockchain555

High

# OptionTokens corresponding to the dust amount of paymentToken cannot be exercised.

## Summary
Due to inaccurate validation in the `OptionTokenV4.sol#_transferRewardToGauge()` function, OptionTokens corresponding to the dust amount of paymentToken cannot be exercised.
## Vulnerability Detail
OptionToken holders can exercise options using the `_exercise()`, `_exerciseVe()` and `_exerciseLp()` functions.

Let’s look at the operation flow of the paymentToken in these function:

Step 1. Here, the user calculates the amount of `paymentToken` to be paid using the `getDiscountedPrice()` and `getPaymentTokenAmountForExerciseLp()` functions.
    https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L579
    https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L607
    https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L666
  These functions use the `getTimeWeightedAveragePrice()` function to calculate the amount of paymentToken.
    https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L324
    https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L335
  The `getTimeWeightedAveragePrice()` function is as follows:
    ```solidity
        function getTimeWeightedAveragePrice(
        uint256 _amount
        ) public view returns (uint256) {
    375:    uint256[] memory amtsOut = IPair(pair).prices(
    376:        underlyingToken,
    377:        _amount,
    378:        twapPoints
    379:    );
            uint256 len = amtsOut.length;
            uint256 summedAmount;

            for (uint256 i = 0; i < len; i++) {
                summedAmount += amtsOut[i];
            }

    387:    return summedAmount / twapPoints;
        }
    ```
  As you can see, the amount of paymentToken is calculated based on the underlying token.
Step 2. Next, the amount of paymentToken compensated by the gauge is calculated in the `_takeFees()` function.
  ```solidity
      function _takeFees(
        address token,
        uint256 paymentAmount
        ) internal returns (uint256 remaining) {
            remaining = paymentAmount;
            for (uint i; i < treasurys.length; i++) {
                uint256 _fee = (paymentAmount * treasurys[i].fee) / 100;
                _safeTransferFrom(token, msg.sender, treasurys[i].treasury, _fee);
        726:    remaining = remaining - _fee;

                if(treasurys[i].notify) {
                    IOptionTokenNotify(treasurys[i].treasury).notify(_fee);
                }
            }
        }
  ```
Step 3. Next, Gauge rewards are notified to the corresponding gauge using the `_transferRewardToGauge()` function.
  ```solidity
    function _transferRewardToGauge() internal {
        uint256 paymentTokenCollectedAmount = IERC20(paymentToken).balanceOf(address(this));

        if(rewardsAddress != address(0)) {
            _safeTransfer(paymentToken,rewardsAddress,paymentTokenCollectedAmount);
        } else {
  745:      uint256 leftRewards = IGaugeV4(gauge).left(paymentToken);
  
  747:      if(paymentTokenCollectedAmount > leftRewards) { // we are sending rewards only if we have more then the current rewards in the gauge
  748:          _safeApprove(paymentToken, gauge, paymentTokenCollectedAmount);
  749:          IGaugeV4(gauge).notifyRewardAmount(paymentToken, paymentTokenCollectedAmount);
  750:      }
        }
    }
  ```

The basic cause of the issue is the lack of validation in #L747 of step 3.
The [`GaugeV4.sol#notifyRewardAmount()`](https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/GaugeV4.sol#L589) function checks whether `rewardRate[token]` is greater than 0.
However, #L747 only checks `paymentTokenCollectedAmount > leftRewards` and does not check whether `(paymentTokenCollectedAmount + leftRewards) / DURATION > 0`.
Therefore, if `(paymentTokenCollectedAmount + leftRewards) / DURATION == 0`, the user cannot exercise the option.

Sherlock ReadMe states that any ERC20 token can be used and that the Option token can be a standard ERC20 token with 18 decimal.
    > Users can create the liquidity pools for any of the ERC20 tokens in permissionless way.

    > Specific contract requirements
    > Option Token contract
    > Standard ERC20 - 18 decimal tokens are only allowed ( rebase, fee on transfer and not 18 decimals tokens are not supported)

The `paymentTokenCollectedAmount` is calculated according to the amount of OptionToken in step 1, so if an ERC20 token with a very small price, such as SHIBE or LADYS, is used as the underlying token, it can be smaller than DURATION (7 days = 604800).
As a result, if the user holds OptionTokens corresponding to the dust amount of paymentToken, `(paymentTokenCollectedAmount + leftRewards) / DURATION == 0`, and therefore the user cannot exercise the option.
## Impact
If the user holds OptionTokens corresponding to the dust amount of paymentToken, he cannot exercise their option.
## Code Snippet
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L739-L752
https://github.com/sherlock-audit/2024-06-velocimeter/blob/main/v4-contracts/contracts/OptionTokenV4.sol#L372-L388
## Tool used

Manual Review

## Recommendation
It is recommended to modify the `_transferRewardToGauge()` function as follows:
```solidity
    function _transferRewardToGauge() internal {
        uint256 paymentTokenCollectedAmount = IERC20(paymentToken).balanceOf(address(this));

        if(rewardsAddress != address(0)) {
            _safeTransfer(paymentToken,rewardsAddress,paymentTokenCollectedAmount);
        } else {
            uint256 leftRewards = IGaugeV4(gauge).left(paymentToken);

---         if(paymentTokenCollectedAmount > leftRewards) { // we are sending rewards only if we have more then the current rewards in the gauge
+++         if(paymentTokenCollectedAmount > leftRewards && paymentTokenCollectedAmount / DURATION > 0) {
                _safeApprove(paymentToken, gauge, paymentTokenCollectedAmount);
                IGaugeV4(gauge).notifyRewardAmount(paymentToken, paymentTokenCollectedAmount);
            }
        }
    }
```

Duplicate of #521

**Comment by blockchain555** [comment]
Escalate
Hi @nevillehuang , You mentioned that in #521 as follows:
>Invalid, this type of weird tokens are not explictly mentioned to be supported as well, so invalid based on sherlock [rules](https://github.com/sherlock-protocol/sherlock-v2-docs/tree/29a019b8690eb01df3587868c4d7eac1e3459bda/audits/judging/judging#vii-list-of-issue-categories-that-are-not-considered-valid). Additionally, there is also always the possibility of assigning a extremely small discount to work around this by the admin.

>Non-Standard tokens: Issues related to tokens with non-standard behaviors, such as weird-tokens are not considered valid by default unless these tokens are explicitly mentioned in the README.

However, in the ReadME, the OptionToken contract stated that only 18 decimal tokens are allowed.
>Option Token contract
Standard ERC20 - 18 decimal tokens are only allowed ( rebase, fee on transfer and not 18 decimals tokens are not supported)

As you can see, it is mentioned that rebase, fee on transfer and not 18 decimals tokens are not supported, but there is no mention that low price 18 decimal tokens are not supported.

In other words, the LADYS token is an 18 decimal token and fully meets the conditions.
https://etherscan.io/token/0x12970E6868f88f6557B76120662c1B3E50A646bf#readContract
Therefore, I think this issue is valid.

**Comment by sherlock-admin3** [comment]
> Escalate
> Hi @nevillehuang , You mentioned that in #521 as follows:
> >Invalid, this type of weird tokens are not explictly mentioned to be supported as well, so invalid based on sherlock [rules](https://github.com/sherlock-protocol/sherlock-v2-docs/tree/29a019b8690eb01df3587868c4d7eac1e3459bda/audits/judging/judging#vii-list-of-issue-categories-that-are-not-considered-valid). Additionally, there is also always the possibility of assigning a extremely small discount to work around this by the admin.
> 
> >Non-Standard tokens: Issues related to tokens with non-standard behaviors, such as weird-tokens are not considered valid by default unless these tokens are explicitly mentioned in the README.
> 
> However, in the ReadME, the OptionToken contract stated that only 18 decimal tokens are allowed.
> >Option Token contract
> Standard ERC20 - 18 decimal tokens are only allowed ( rebase, fee on transfer and not 18 decimals tokens are not supported)
> 
> As you can see, it is mentioned that rebase, fee on transfer and not 18 decimals tokens are not supported, but there is no mention that low price 18 decimal tokens are not supported.
> 
> In other words, the LADYS token is an 18 decimal token and fully meets the conditions.
> https://etherscan.io/token/0x12970E6868f88f6557B76120662c1B3E50A646bf#readContract
> Therefore, I think this issue is valid.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by nevillehuang** [comment]
This issue lacks a concrete numerical PoC/example that is required from [sherlock for rounding/precision loss issues](https://github.com/sherlock-protocol/sherlock-v2-docs/tree/29a019b8690eb01df3587868c4d7eac1e3459bda/audits/judging/judging#vi-requirements). Additionally, in option contracts, only 18 decimal tokens are allowed

> Specific contract requirements Option Token contract Standard ERC20 - 18 decimal tokens are only allowed ( rebase, fee on transfer and not 18 decimals tokens are not supported)

**Comment by blockchain555** [comment]
> This issue lacks a concrete numerical PoC/example that is required from [sherlock for rounding/precision loss issues](https://github.com/sherlock-protocol/sherlock-v2-docs/tree/29a019b8690eb01df3587868c4d7eac1e3459bda/audits/judging/judging#vi-requirements). Additionally, in option contracts, only 18 decimal tokens are allowed
> 
> > Specific contract requirements Option Token contract Standard ERC20 - 18 decimal tokens are only allowed ( rebase, fee on transfer and not 18 decimals tokens are not supported)

You're right. but low price token is also 18 decimal token.

**Comment by cvetanovv** [comment]
@blockchain555 In the Readme is written:

> **Standard ERC20** - 18 decimal tokens are only allowed 

Although you gave the example of an 18-decimal token, it is clearly written that they will use only **standard** tokens. I don't think the above token is a standard.

Planning to reject the escalation and leave the issue as is.

**Comment by WangSecurity** [comment]
Result:
Invalid
Duplicate of #521

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [blockchain555](https://github.com/sherlock-audit/2024-06-velocimeter-judging/issues/176/#issuecomment-2283479540): rejected
